pkg/solana/txm/txm.go

```diff
@@ -377,139 +399,209 @@ func (txm *Txm) confirm() {
+// It updates the transaction state in the local memory and checks if the confirmation timeout has been exceeded.
+// If the timeout is exceeded, it marks the transaction as errored.
+func (txm *Txm) handleProcessedSignatureStatus(sig solanaGo.Signature) {
+	// update transaction state in local memory
+	id, err := txm.txs.OnProcessed(sig)
+	if err != nil && !errors.Is(err, ErrAlreadyInExpectedState) {
+		txm.lggr.Errorw("failed to mark transaction as processed", "signature", sig, "error", err)
+	} else if err == nil {
+		txm.lggr.Debugw("marking transaction as processed", "id", id, "signature", sig)
+	}
+	// check confirm timeout exceeded if TxConfirmTimeout set
+	if txm.cfg.TxConfirmTimeout() != 0*time.Second && txm.txs.Expired(sig, txm.cfg.TxConfirmTimeout()) {
+		id, err := txm.txs.OnError(sig, txm.cfg.TxRetentionTimeout(), Errored, TxFailDrop)
+		if err != nil {
+			txm.lggr.Infow("failed to mark transaction as errored", "id", id, "signature", sig, "timeoutSeconds", txm.cfg.TxConfirmTimeout(), "error", err)
+		} else {
+			txm.lggr.Debugw("tx failed to move beyond 'processed' within confirm timeout", "id", id, "signature", sig, "timeoutSeconds", txm.cfg.TxConfirmTimeout())
 		}
-		tick = time.After(utils.WithJitter(txm.cfg.ConfirmPollPeriod()))
+	}
+}
+
+// handleConfirmedSignatureStatus handles the case where a transaction signature is in the "confirmed" state on-chain.
+// It updates the transaction state in the local memory.
+func (txm *Txm) handleConfirmedSignatureStatus(sig solanaGo.Signature) {
+	id, err := txm.txs.OnConfirmed(sig)
+	if err != nil && !errors.Is(err, ErrAlreadyInExpectedState) {
+		txm.lggr.Errorw("failed to mark transaction as confirmed", "id", id, "signature", sig, "error", err)
+	} else if err == nil {
+		txm.lggr.Debugw("marking transaction as confirmed", "id", id, "signature", sig)
+	}
+}
+
+// handleFinalizedSignatureStatus handles the case where a transaction signature is in the "finalized" state on-chain.
+// It updates the transaction state in the local memory.
+func (txm *Txm) handleFinalizedSignatureStatus(sig solanaGo.Signature) {
+	id, err := txm.txs.OnFinalized(sig, txm.cfg.TxRetentionTimeout())
+	if err != nil {
+		txm.lggr.Errorw("failed to mark transaction as finalized", "id", id, "signature", sig, "error", err)
+	} else {
+		txm.lggr.Debugw("marking transaction as finalized", "id", id, "signature", sig)
+	}
+}
+
+// rebroadcastExpiredTxs attempts to rebroadcast all transactions that are in broadcasted state and have expired.
+// An expired tx is one where it's blockhash lastValidBlockHeight is smaller than the current slot height.
+// If any error occurs during rebroadcast attempt, they are discarded, and the function continues with the next transaction.
+func (txm *Txm) rebroadcastExpiredTxs(ctx context.Context, client client.ReaderWriter) {
+	currBlockHeight, err := client.GetLatestBlock(ctx)
```

Comment:
nit: this is probably a misnomer for this variable. Think we should call this something like `currBlock` or `latestBlock` since that's what's really being returned by the RPC call

Reply:
Thanks for this! Fixed